test/test_sparse_csr.py

```diff
@@ -1409,6 +1409,18 @@ def test_exercise_detach(self, device, dtype):
             self.assertEqual(inp.crow_indices(), detached_inp.crow_indices())
             self.assertEqual(inp.col_indices(), detached_inp.col_indices())
 
+    @skipMeta
+    @dtypes(*get_all_dtypes(include_bfloat16=False, include_complex=False))
```

Comment:
Do tests pass for `include_bfloat16=True` and `include_complex=True`?

Reply:
Thanks, @IvanYashchuk for asking this question. It only fails for `bfloat16`, I've revised the types in the latest comment. Also for Sparse COO, `eye` supports `complex64, complex128` but not `complex32`, so I've revised that as well.

Reply:
What's the error message for `bfloat16`?

Reply:
From what observed, the reasons are 2:

1. `eye` not implemented for `BFloat16`
2. `to_dense()` doesn't work with `BFloat16` (for Sparse COO)

Following is the error that I'm getting:

```python
======================================================================
ERROR: test_eye_cpu_bfloat16 (__main__.TestSparseCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/krshrimali/pytorch/test/test_sparse.py", line 3412, in test_eye
    dense_output = torch.eye(n, device=device, dtype=dtype)
RuntimeError: "eye" not implemented for 'BFloat16'
```

Reply:
It works on CUDA.
```py
In [1]: import torch

In [2]: torch.eye(3, dtype=torch.bfloat16)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-2-27c993df381a> in <module>
----> 1 torch.eye(3, dtype=torch.bfloat16)

RuntimeError: "eye" not implemented for 'BFloat16'

In [3]: torch.eye(3, dtype=torch.bfloat16, device='cuda')
Out[3]:
tensor([[1., 0., 0.],
        [0., 1., 0.],
        [0., 0., 1.]], device='cuda:0', dtype=torch.bfloat16)

In [4]: torch.eye(3, dtype=torch.bfloat16, device='cuda').to_sparse()
Out[4]:
tensor(indices=tensor([[0, 1, 2],
                       [0, 1, 2]]),
       values=tensor([1., 1., 1.]),
       device='cuda:0', size=(3, 3), nnz=3, dtype=torch.bfloat16,
       layout=torch.sparse_coo)

In [5]: torch.eye(3, dtype=torch.bfloat16, device='cuda').to_sparse().to_dense()
Out[5]:
tensor([[1., 0., 0.],
        [0., 1., 0.],
        [0., 0., 1.]], device='cuda:0', dtype=torch.bfloat16)
```

Reply:
Enabling `eye` for bfloat16 on CPU shouldn't be much of an effort. Probably requires just changing the dispatch macro.

Reply:
Right, but then `to_dense()` will still fail for `BFloat16` types on CPU.

```python
ERROR: test_eye_cpu_bfloat16 (__main__.TestSparseCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/krshrimali/pytorch/torch/testing/_internal/common_device_type.py", line 389, in instantiated_test
    raise rte
  File "/home/krshrimali/pytorch/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
    result = test(self, **param_kwargs)
  File "/home/krshrimali/pytorch/torch/testing/_internal/common_device_type.py", line 796, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/home/krshrimali/pytorch/test/test_sparse.py", line 3414, in test_eye
    self.assertEqual(sparse_coo_output.to_dense(), dense_output)
RuntimeError: "add_dense_sparse" not implemented for 'BFloat16'
```

Locally, in the `add_out_sparse_cpu` function in `SparseTensorMath.cpp` file, if I change the dispatch macro to add `BFloat16`, the test passes. I'm pushing these commits, if you can take a look - that will help!

Reply:
> Locally, in the `add_out_sparse_cpu` function in `SparseTensorMath.cpp` file, if I change the dispatch macro to add `BFloat16`, the test passes. I'm pushing these commits, if you can take a look - that will help!

I tried adding `Half` in the dispatch macro as well, and the test passes with `Half` as a dtype. Should we do that? Or was there a reason that it wasn't done before?

Reply:
This should be ready for review now :)